### PR TITLE
Fixes: 4589 - "enhancement: more descriptive error for a particular l…

### LINKED ISSUE #4589: enhancement: more descriptive error for a particular loadonly usage
The error message on this tripped me up for a bit until I realized what happened. It would be nice if SqlAlchemy could raise something a bit more descriptive of the situation...

A project recently migrated some of it's schema, and a handful of columns were made readable via the legacy interface using standard Python `@property` decorator.

I missed migrating some of the `load_only` elements, and the error raised by SqlAlchemy is:

   AttributeError: 'property' object has no attribute 'property'

While the error makes perfect sense once you know what caused it, debugging it from scratch was a pain as you can't do much with a standard property object in pdb.  It took a long time to figure out the error was from providing a bad column name to loadonly. 

It would be great to catch this issue and state something like 'The column you are attempted to load is a @property object and not an actual column`. 

i don't think doing so would create issues elsewhere, but wanted to propose this enhancement for feedback before attempting a PR.

the attached code illustrates this.  

```
Traceback (most recent call last):
  File "/bugreport/test_orm_changed.py", line 73, in <module>
    print s.query(Foo).options(joinedload('bar').load_only('biz')).first()
  File "/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 1358, in options
    return self._options(False, *args)
  File "<string>", line 2, in _options
  File "/lib/python2.7/site-packages/sqlalchemy/orm/base.py", line 201, in generate
    fn(self, *args[1:], **kw)
  File "/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 1375, in _options
    opt.process_query(self)
  File "/lib/python2.7/site-packages/sqlalchemy/orm/strategy_options.py", line 146, in process_query
    self._process(query, True)
  File "/lib/python2.7/site-packages/sqlalchemy/orm/strategy_options.py", line 468, in _process
    query._current_path, query._attributes, raiseerr)
  File "/lib/python2.7/site-packages/sqlalchemy/orm/strategy_options.py", line 639, in _bind_loader
    loader.path, token, None, raiseerr):
  File "/lib/python2.7/site-packages/sqlalchemy/orm/strategy_options.py", line 200, in _generate_path
    attr = attr.property
AttributeError: 'property' object has no attribute 'property'
```

[test_orm_changed.txt](https://github.com/sqlalchemy/sqlalchemy/files/3040778/test_orm_changed.txt)

### PR DESCRIPTION
**Pull request by jvanasco**
Fixes: #4589 - "enhancement: more descriptive error for a particular loadonly usage"

* Detect when a class attribute passed to `load_only(` is not actually a column, but a Python class's method that is decorated with `@property`; raise a more descriptive error.
* Test case for raising the correct exception and message when a Python @property column is submitted to `load_only()`

### Description

`_generate_path` now catches an exception, looking for a particular situation where the attribute is a Python @property

A new test case `InvalidLoadOnlyTest` was added to `test/orm/test_attributes.py`.  It was modeled after `test/orm/inheritance/test_basic.py` but did not fit in the inheritance section. Because of this, several support functions are newly imported into `test_attributes.py`; instead of integrating them into the list of all imports, they appear as a separate section in case this is best moved to another test suite.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
- [X] A new feature implementation
        - Fixes #4589
        - Tests are included

### CODE REVIEW

File: lib/sqlalchemy/orm/strategy_options.py [hunk 1]
```diff
@@ -232,7 +232,16 @@ def _generate_path(
                 else:
                     return None
             else:
-                attr = found_property = attr.property
```
Comment (zzzeek) [inline lib/sqlalchemy/orm/strategy_options.py]: OK, how about, "if not isinstance(attr, MapperProperty):  raise ...("Expected %s to be a mapped column; instaed got %r object")

that way if they have any other kind of attribute or descriptor there we still raise something meaningful

Comment (jvanasco) [reply lib/sqlalchemy/orm/strategy_options.py]: that's a much better idea.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 2b7ed5240f49be90f9390e3d041c9cb957083465 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 2b7ed5240f49be90f9390e3d041c9cb957083465: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1908

**Comment by zzzeek** [comment]
wow it branched cleanly

**Comment by jvanasco** [comment]
The pep8 errors confuses me - black in the pre-commit should have handled that. This PR could have been against an earlier version of black though.  

The test errors seem to be because `<type 'property'>` vs `<class 'property'>`.  Could this be dependent on the platform, or did I just typo that?

**Comment by zzzeek** [comment]
its likely not a typo, it's trying to be intelligent inside of a `__getattr__` loop.  these are the most difficult things to debug in Python so I don't really know what the problem is.  messing with the getattr stuff is very risky as it often leads to regressions with people that have hybrid attributes to funny objects and things like that.

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1908

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2182

**Comment by sqla-tester** [comment]
**Jonathan Vanasco** (jvanasco) wrote:

Wow, this required a lot of changes. I'm sorry.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1908

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1908

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2182

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+2

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1908

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1908 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+2

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2182

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2182 has been **merged**. Congratulations! :)

**Pull request closed without merging**